### PR TITLE
19 implementing viewing functionality for startup profiles

### PR DESCRIPTION
**Pull request by AnnKvytko**
This pull request implements the viewing functionality for startup profiles, allowing investors to access detailed information about various startups. The following enhancements and modifications have been made:

- Modified existing **endpoints** to facilitate the retrieval of startup profile information.
- Developed custom **permissions** to ensure that only authorized users with an active investor role have access to retrieve startup profile information.
- Implemented **search** functionality that allows investors to search for startups by name, enhancing the user experience.
- Added **filtering** capabilities based on criteria such as industry and name.
- Developed a comprehensive view for each startup profile, providing investors with information about the startups they are interested in.
- Wrote tests to ensure the profile viewing feature functions correctly.
requests
postman2
postman5
postman4
postman3

### CODE REVIEW

File: forum/startups/tests.py [hunk 1]
```diff
@@ -217,5 +221,109 @@ def test_profile_creation(self):
...
+            user_id=self.user.user_id,
+            company_name="Solar",
+            required_funding=1000000,
+            location=self.location,
+        )
+        self.startup.industries.set([self.industry1])
+        self.startup2.industries.set([self.industry2])
+
+        self.list_url = reverse('startup-list')
+        self.retrieve_url = reverse('startup-detail', args=[self.startup.startup_id])
```
Comment (mehalyna) [inline forum/startups/tests.py:287]: After creating startup instances, include assertions to confirm these objects were created successfully. This will ensure `setUp` ran as expected and help with debugging if an error occurs here.
```
def setUp(self):
    # Existing setup code...
    self.assertTrue(Startup.objects.filter(company_name="EcoSolutionss").exists())
    self.assertTrue(Startup.objects.filter(company_name="Solar").exists())
```

File: forum/startups/tests.py
Comment (mehalyna) [inline forum/startups/tests.py:1]: Add a test to ensure the API returns an empty list if no startups match the filter criteria. This is helpful to check the behavior when no records are found.
```
def test_list_startups_no_results(self):
    response = self.client.get(self.list_url, {'search': 'NonExistentCompany'})
    self.assertEqual(response.status_code, status.HTTP_200_OK)
    self.assertEqual(len(response.data), 0)
```

Comment (mehalyna) [inline forum/startups/tests.py:1]: Since this test suite involves an investor role, add tests to verify access for users without the appropriate role. You could mock an authenticated user with a different role, such as startup, and ensure they receive a `403 Forbidden` response.
```
def test_access_denied_for_non_investor(self):
    self.user.change_active_role('startup')
    response = self.client.get(self.list_url)
    self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
```

File: forum/startups/views.py [hunk 2]
```diff
@@ -101,6 +147,90 @@ def destroy(self, request, *args, **kwargs):
...
+            queryset = self.get_queryset()
+            serializer = self.get_serializer(queryset, many=True)
+            return Response(serializer.data)
+        except ValidationError as e:
+            return Response({"detail": str(e)}, status=status.HTTP_400_BAD_REQUEST)
+        except IntegrityError:
+            return Response({"detail": "Database integrity error occurred."},
+                        status=status.HTTP_400_BAD_REQUEST)
+        except Exception as e:
+            return Response({"detail": str(e)}, status=status.HTTP_500_INTERNAL_SERVER_ERROR)
```
Comment (mehalyna) [inline forum/startups/views.py:200]: The method currently fetches all startups, which may lead to performance issues when dealing with a large dataset.
Integrate `PageNumberPagination` in `list` to return a subset of records per page, optimizing data handling and response time.
```
from rest_framework.pagination import PageNumberPagination

class StandardResultsSetPagination(PageNumberPagination):
    page_size = 10
    page_size_query_param = 'page_size'
    max_page_size = 100

class StartupViewSet(viewsets.ModelViewSet):
    # other code ...

    def list(self, request, *args, **kwargs):
        paginator = StandardResultsSetPagination()
        queryset = self.get_queryset()
        paginated_queryset = paginator.paginate_queryset(queryset, request)
        serializer = self.get_serializer(paginated_queryset, many=True)
        return paginator.get_paginated_response(serializer.data)
```

File: forum/startups/views.py
Comment (mehalyna) [inline forum/startups/views.py:1]: Permissions are checked at multiple places (in `get_permission`s and `get_queryset`), which could be simplified.
Instead of directly checking roles in `get_queryset`, define permissions in the viewset’s `permission_classes` with custom permissions, ensuring consistent access control.

File: forum/startups/views.py [hunk 3]
```diff
@@ -101,6 +147,90 @@ def destroy(self, request, *args, **kwargs):
...
+        company_name = self.request.query_params.get('company_name', None)
+        industry = self.request.query_params.get('industry', None)
+
+        if company_name:
+            queryset = queryset.filter(company_name__icontains=company_name)
+        industry_name = self.request.query_params.get('industry_name', None)
+        if industry_name:
+            queryset = queryset.filter(industries__name__icontains=industry_name)
+
+        return queryset
```
Comment (mehalyna) [inline forum/startups/views.py:232]: `get_queryset` has multiple checks for different query parameters, making the method lengthy.
Separate filtering logic into a dedicated helper method for readability and scalability.
```
def filter_queryset(self, queryset):
    search = self.request.query_params.get('search')
    company_name = self.request.query_params.get('company_name')
    industry_name = self.request.query_params.get('industry_name')

    if search:
        queryset = queryset.filter(company_name__icontains=search)
    if company_name:
        queryset = queryset.filter(company_name__icontains=company_name)
    if industry_name:
        queryset = queryset.filter(industries__name__icontains=industry_name)

    return queryset

def get_queryset(self):
    queryset = super().get_queryset()
    return self.filter_queryset(queryset)
```

File: forum/startups/views.py
Comment (mehalyna) [inline forum/startups/views.py:1]: Only the logger object is declared, but it’s not utilized in methods like `retrieve` and `list`.
Add logging statements for debugging and tracking the API’s performance and request details.

**Review by mehalyna — CHANGES_REQUESTED** [review]

**Review by mehalyna — APPROVED** [review]

**Review by k0drin — APPROVED** [review]
